### PR TITLE
mds: trim cache during standby-replay

### PR DESCRIPTION
**Pull request by simon-rock**
Fixes:http://tracker.ceph.com/issues/40213
Signed-off-by: simon gao <simon29rock@gmail.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -6522,6 +6522,7 @@ std::pair<bool, uint64_t> MDCache::trim_lru(uint64_t count, expiremap& expiremap
     if ((is_standby_replay && dn->get_linkage()->inode &&
         dn->get_linkage()->inode->item_open_file.is_on_list())) {
       unexpirables.push_back(dn);
```
Comment (ukernel) [inline src/mds/MDCache.cc]: call lru.lru_insert_bot(dn) here. it's more readable

File: src/mds/MDCache.cc [hunk 2]
```diff
@@ -6512,7 +6512,8 @@ std::pair<bool, uint64_t> MDCache::trim_lru(uint64_t count, expiremap& expiremap
   unexpirables.clear();
 
   // trim dentries from the LRU until count is reached
-  while (!throttled && (cache_toofull() || count > 0)) {
+  // if mds is in standbyreplay and will trim all inodes which aren't in segments
```
Comment (batrick) [inline src/mds/MDCache.cc]: :O Why do we want to trim dentries not in a log segment? Wouldn't this prevent the standby-replay MDS from keeping its cache hot with all recently used metadata?

File: src/mds/MDCache.cc [hunk 3]
```diff
@@ -6521,7 +6522,10 @@ std::pair<bool, uint64_t> MDCache::trim_lru(uint64_t count, expiremap& expiremap
     }
     if ((is_standby_replay && dn->get_linkage()->inode &&
         dn->get_linkage()->inode->item_open_file.is_on_list())) {
-      unexpirables.push_back(dn);
+      // we move the inodes that need to be trimmed to the end of the lru queue.
+      // refer to MDCache::standby_trim_segment
+      lru.lru_insert_bot(dn);
+      break;
```
Comment (batrick) [inline src/mds/MDCache.cc]: This presents the possibility of an infinite loop. I don't really see the need for this change.

Comment (ukernel) [reply src/mds/MDCache.cc]: it just inserts dn back and break the loop

File: src/mds/MDCache.cc (hunk 2 above)
Comment (ukernel) [reply src/mds/MDCache.cc]: standby replay mds only knows which metadata are recently modified.  recently modified metatdata is not equivalent to hot metadata

**Comment by ukernel** [comment]
yes, standby replay mds does not trim its cache

**Pull request merged**